### ISSUE TITLE
Question Wildcards

### ISSUE THREAD

**Issue by crbyxwpzfl** [theelims/PsychicMqttClient#6]
Hi thank you so much for this lib!

**tldr;** how do I use mqtt [wildcards](https://www.emqx.com/en/blog/advanced-features-of-mqtt-topics#mqtt-wildcards)?  In the code wildcard support is mentioned [here](https://github.com/theelims/PsychicMqttClient/blob/ede67b47f484d37a71c9e49f1b599ae60f802d9b/src/PsychicMqttClient.h#L251) so I just tried the [examples/Simple/main.cpp](https://github.com/theelims/PsychicMqttClient/blob/ede67b47f484d37a71c9e49f1b599ae60f802d9b/examples/Simple/main.cpp)
with `"/2098345798037245/#";` and thought this will subscribe to all subtopics. But I do not recive any messages with this. 

Please close this whenever. this issue likely is me not reading the docs correctly or misunderstanding something. anyways perhaps you have some tips for me.

**more info** I tried `String subtopic = "/2098345798037245/+";` as well but I did not receive any messages either.
with `mqttClient.onSubscribe(onMqttSubscribe);` I do get `Subscribe acknowledged.` with a wildcard in the topic.
when I add both `mqttClient.onTopic( "/2098345798037245/test" ...` and `mqttClient.onTopic( "/2098345798037245/#" ...` the lambda without the wildcard seems to be called twice when it receives one message on `/2098345798037245/test`.
I tested this with [this web client](https://mqttx.app/web-client) and [this](https://www.emqx.com/en/mqtt/public-mqtt5-broker) public broker.

<details>
<summary>full code and output</summary>
<br>

```c

#include <Arduino.h>
#include <WiFi.h>
#include <PsychicMqttClient.h>

const char ssid[] = "ssid"; // your network SSID (name)
const char pass[] = "pass"; // your network password

PsychicMqttClient mqttClient;

void onMqttSubscribe(uint16_t packetId)
{
    Serial.println("Subscribe acknowledged.");
    Serial.printf("  packetId: %d\r\n", packetId);
}

void setup()
{
    Serial.begin(115200);

    WiFi.begin(ssid, pass);

    Serial.printf("Connecting to WiFi %s .", ssid);
    while (WiFi.status() != WL_CONNECTED)
    {
        Serial.print(".");
        delay(500);
    }

    Serial.printf("\r\nConnected, IP address: %s \r\n", WiFi.localIP().toString().c_str());

    mqttClient.setServer("mqtt://broker.emqx.io");

    mqttClient.onSubscribe(onMqttSubscribe);

    String topic = "/2098345798037245/1";

    mqttClient.onTopic("/2098345798037245/test", 0, [&](const char *topic, const char *payload, int retain, int qos, bool dup)
                       { Serial.printf("NO WILDCARD Received Topic: %s\r\n", topic);
                         Serial.printf("NO WILDCARD Received Payload: %s\r\n", payload); });

    mqttClient.onTopic("/2098345798037245/#", 0, [&](const char *topic, const char *payload, int retain, int qos, bool dup)
                       { Serial.printf("Received Topic: %s\r\n", topic);
                         Serial.printf("Received Payload: %s\r\n", payload); });

    mqttClient.connect();

    while (!mqttClient.connected())
    {
        delay(500);
    }

    mqttClient.publish(topic.c_str(), 0, 0, "Hello World!");
}

void loop() {}

```

output 
```
Connected, IP address: 192.168.178.119 
Subscribe acknowledged.
  packetId: 32720
Subscribe acknowledged.
  packetId: 3923

NO WILDCARD Received Topic: /2098345798037245/test
NO WILDCARD Received Payload: This is published just once
NO WILDCARD Received Topic: /2098345798037245/test
NO WILDCARD Received Payload: This is published just once
```

</details>

**Comment by theelims** [comment]
Hmm, I did test wildcards, although not intensely. But your example suggests there is something off. I need to dig into it. Unfortunately life is quite busy at the moment. I am not sure when I'll find the time to do so.

**Comment by crbyxwpzfl** [comment]
thank you for your quick reply. I totally understand plus this issue is not at all important. It is easy to just manually subscribe to the topics instead. ttyl, cheers

**Comment by theelims** [comment]
The function in question was created by Copilot. It is no surprise that it is doing only 80% of what it is supposed to do. 😆

**Comment by sluebbert** [comment]
I got hit by this as well and I think it comes down to the `_isTopicMatch` method.

For example, the topic is:
`devices/EM1/oMaRjWZ554jdMf2Y/0/ref`
and the subscription is:
`devices/EM1/oMaRjWZ554jdMf2Y/+/ref`

The `strtok` usage looks like a culprit. Docs point to not using it for two different operations at the same time since it saves state internally. I think using it for slicing up the topic string and the subscription string at the same time is causing a collision.

It probably works just fine in the first iteration for the loop for short subscriptions, but for me, logging shows odd outcomes for the next iterations.

**Comment by crbyxwpzfl** [comment]
hi this seems promising I don’t know much about c. I will try to read into it a bit and try to understand strtok. Thank you for that you are still interested in this!

**Comment by theelims** [comment]
@sluebbert @crbyxwpzfl Could you please test the latest code in the main repository? I refactored the wildcard matching function and in my tests it works. See the Wildcard example.

**Comment by sluebbert** [comment]
I can't say I thoroughly tested the subscription matching changes, but my use case now works as expected :slightly_smiling_face:

Thanks for the help!

**Comment by crbyxwpzfl** [comment]
thank you @theelims! Unfortunately i will not be able to test for the next week im really sorry! But this is exiting to me and I will test as early as possible!
